rfc/src/rfcs/0005-should-fail-attr.md

```diff
@@ -0,0 +1,197 @@
+Negative verification harnesses could be useful to show that invalid inputs result in verification failures,
+or increase the overall verification coverage.
+
+In fact, some of the tests in the Kani regression are written as negative harnesses.
+Currently, we [use comments](https://model-checking.github.io/kani/regression-testing.html#testing-stages) specific to `compiletest` to indicate that a verification failure is expected.
+The proposed attribute would provide a way in Kani to express that a verification harness should fail,
+allowing users to write negative verification harnesses.
+
+We also acknowledge that, in other cases, users may want to express more granular expectations for their harnesses.
+For example, a user may want to specify that a given check results in a failure.
+An ergonomic mechanism for informing Kani about such expectations is likely to require other improvements in Kani (a comprehensive classification for checks reported by Kani, a language to describe expectations for checks and cover statements, and general output improvements).
+We consider that the mechanism we just mentioned and the one discussed in this proposal solve different problems, so they don't need to be discussed together.
+This is further discussed in the [rationale and alternatives](#rationale-and-alternatives) and [future possibilities](#future-possibilities) sections.
+
+## User Experience
+
+The scope of this functionality is limited to the overall verification result.
+The [rationale section](#rationale-and-alternatives) discusses the granularity of failures, and how this attribute could be extended.
+
+### Single Harness
+
+Let's take an example from the Kani regression:
+
+```rust
+#[kani::proof]
+fn simple_add_overflows() {
+    let a: u32 = kani::any();
+    let b: u32 = kani::any();
+    a + b;
+}
+```
+
+Currently, this example produces a `VERIFICATION:- FAILED` result.[^footnote-compiletest]
+In addition, it will return a non-successful code.
+
+```rust
+#[kani::proof]
+#[kani::should_fail]
+fn simple_add_overflows() {
+    let a: u32 = kani::any();
+    let b: u32 = kani::any();
+    a + b;
+}
+```
+
+Since we added `#[kani::should_fail]`, running this example would produce a successful code.
+
+Now, we've considered two ways to represent this result in the verification output.
+Note that it's important that we provide the user with this feedback:
+ 1. **(Expectation)** Kani was expecting the harness to fail.
+ 2. **(Outcome)**: The actual verification result that Kani produced after the analysis.
+This will avoid a potential scenario where the user doesn't know for sure if the attribute has had an effect when running the verification harness.
+
+As mentioned, we've considered two ways to represent this result.
+
+#### Representation #1 (Recommended): No changes to overall result
+
+```rust
+VERIFICATION:- FAILED (expected FAILED)
```

Comment:
I'm so glad you are tackling this here, but I wonder if we can still improve this. Double negation is hard to parse.

Reply:
True, it's also not clear to me how this result will look in #1690. I need to spend more time on this one.

Reply:
Couldn't find a simpler representation. It's now a question in the "Open questions", and I'll explicitly look for feedback on this.